### ISSUE TITLE
How to run function after document's readyState?

### ISSUE THREAD

**Issue by timfayz** [gopherjs/gopherjs#316]
Hello guys) How can I implement this code in Go?

```JavaScript
document.onreadystatechange = function () {
    console.log("test");
}
```

I've tried the beyond, but it doesn't work..

```Go
package main

import "github.com/gopherjs/gopherjs/js"

func main() {
  js.Global.Get("document").Set("onreadystatechange", func() {
	print("test") // I think "test" must be fired several times, but it just output nothing
  })
}
```

**Comment by neelance** [comment]
I just tested this code and it works fine to me. Which browser do you use?

**Comment by timfayz** [comment]
Good!) The problem was folly: I placed the js code I want to implement after the code was generated by gopher, so it rewrite `onreadystatechange` property. That's all!)

In order not to create plenty of issues I want to ask additional questions here. For example, quote from readme:

> Goroutines are fully supported by GopherJS. The only restriction is that you need to start a new goroutine if you want to use blocking code called from external JavaScript:

```
js.Global.Get("myButton").Call("addEventListener", "click", func() {
  go func() {
    [...]
    someBlockingFunction()
    [...]
  }()
})
```
Could you give some concrete example, please? For example, a situation where I must to do that and must not.

**Comment by shurcooL** [comment]
Hi @timfayz,

I'll answer your question with a very simple example. I'll use [`dom` package](https://godoc.org/honnef.co/go/js/dom) in my code so the add even listener will look slightly different.

```Go
textArea := document.GetElementByID("some-textarea-element").(*dom.HTMLTextAreaElement)

textArea.AddEventListener("click", false, func(e dom.Event) {
	fmt.Println("Clicked right away.")
})
```

That will work okay because the call to `fmt.Println` is not blocking, it can run without any blocking I/O, no waiting for channels, etc.

Let's change it to be blocking:

```Go
textArea.AddEventListener("click", false, func(e dom.Event) {
	time.Sleep(time.Second)
	fmt.Println("Clicked after blocking for a second.")
})
```

(It can be other blocking code like making XHR request and waiting for result, receiving on a channel, etc.)

You'll get the following error in console:

![image](https://cloud.githubusercontent.com/assets/1924134/10262186/956d81f0-6971-11e5-81bf-5d6d8eddfa33.png)

Let's fix it by wrapping the code in a background goroutine so that the JavaScript callback (for "click" event) is not blocking and can return quickly:

```Go
textArea.AddEventListener("click", false, func(e dom.Event) {
	go func() {
		time.Sleep(time.Second)
		fmt.Println("Clicked after blocking for a second.")
	}()
})
```

That fixes the issue.

In summary, IMO the best practice is to start simple and just write your non-blocking code in a JS callback handler. But as soon as your code becomes blocking (as you'll see from the above error message), then make it non-blocking by putting the blocking code inside `go func() { ... }()`.

Does that help?

**Comment by timfayz** [comment]
Yep! It's more than enough. Thank you)
